consensus/src/experimental/commit_phase.rs

```diff
@@ -0,0 +1,333 @@
+            return Ok(()); // we ignore the message
+        }
+
+        // verify the signature
+        commit_vote.verify(&self.verifier)?;
+
+        let chash = li.commit_info().executed_state_id();
+
+        // add the signature to local_cache
+        match self.local_cache.entry(chash) {
+            Entry::Occupied(mut cio) => {
+                let ci = cio.get_mut();
+                ci.add_signature(commit_vote.author(), commit_vote.signature().clone());
+            }
+            Entry::Vacant(_) => {
+                let mut li_sig = LedgerInfoWithSignatures::new(
+                    li.clone(),
+                    BTreeMap::<AccountAddress, Ed25519Signature>::new(),
+                );
+                li_sig.add_signature(commit_vote.author(), commit_vote.signature().clone());
+                self.local_cache.insert(chash, li_sig);
+            }
+        };
+
+        Ok(())
+    }
+
+    pub async fn process_commit_decision(
+        &mut self,
+        commit_decision: CommitDecision,
+    ) -> anyhow::Result<()> {
+        let li = commit_decision.ledger_info();
+
+        if li.ledger_info().commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        commit_decision.verify(&self.verifier)?;
+
+        let chash = li.ledger_info().commit_info().executed_state_id();
+
+        // TODO: optimization1: probe local_cache first to see if the existing already verifies,
+        // TODO: otherwise we do not make changes.
+        // TODO: optimization2: we can set a bit to indicate this tree of signatures are already verified,
+        // TODO: we do not have to verify it again in the main loop.
+
+        // replace the signature tree directly if it does not verify
+        self.local_cache.insert(chash, li.clone());
+
+        Ok(())
+    }
+
+    pub async fn start(mut self) {
+        loop {
+            select! {
+                (vecblock, ledger_info) = self.commit_channel_recv.select_next_some() => {
+                    let new_ledger_info = LedgerInfo::new(
+                        vecblock.last().unwrap().block_info(),
+                        ledger_info.ledger_info().consensus_data_hash(),
+                    );
+
+                    let sig = self.safety_rules
+                        .lock()
+                        .sign_commit_vote(ledger_info.clone(), new_ledger_info.clone()).map_err(|err| {
+                            counters::ERROR_COUNT.inc();
+                            error!(
+                                epoch = new_ledger_info.commit_info().epoch(),
+                                error = err,
+                                "Error occurred in signing commit vote."
+                            )
+                        }).unwrap();
+
+                    self.commit_msg_sender.send(
+                        CommitPhaseChannelMsgWrapper::CommitVote(self.author, new_ledger_info.clone(), sig)
```

Comment:
need to think about unhappy path here too (not in this PR), what if the message is lost

Reply:
Got it. I'll add a footnote